src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/BoxingHelpers.cs

```diff
@@ -0,0 +1,124 @@
+        {
+            uint numInstanceFieldBytes = pMT->GetNumInstanceFieldBytes();
+            if (((uint)Unsafe.AsPointer(ref destBytes) | numInstanceFieldBytes & ((uint)sizeof(void*) - 1)) != 0)
+            {
+                // If we have a non-pointer aligned instance field bytes count, or a non-aligned destBytes, we can zero out the data byte by byte
+                // And we do not need to concern ourselves with references
+                SpanHelpers.ClearWithoutReferences(ref destBytes, numInstanceFieldBytes);
+            }
+            else
+            {
+                // Otherwise, use the helper which is safe for that situation
+                SpanHelpers.ClearWithReferences(ref Unsafe.As<byte, IntPtr>(ref destBytes), (nuint)numInstanceFieldBytes / (nuint)sizeof(IntPtr));
+            }
+        }
+
+        [DebuggerHidden]
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static bool AreTypesEquivalent(MethodTable* pMTa, MethodTable* pMTb)
+        {
+            if (pMTa == pMTb)
+            {
+                return true;
+            }
+
+            if (pMTa->HasTypeEquivalence && pMTb->HasTypeEquivalence)
+            {
+                return false;
+            }
+
+            return RuntimeHelpers.AreTypesEquivalent(pMTa, pMTb);
+        }
+
+        [DebuggerHidden]
+        private static bool IsNullableForType(MethodTable* typeMT, MethodTable* boxedMT)
+        {
+            if (!typeMT->IsNullable)
+            {
+                return false;
+            }
+
+            MethodTable *pMTNullableArg = typeMT->InstantiationArg0();
+            if (pMTNullableArg == boxedMT)
+            {
+                return true;
+            }
+            else
+            {
+                return AreTypesEquivalent(pMTNullableArg, boxedMT);
+            }
+        }
+
+        [DebuggerHidden]
+        internal static void Unbox_Nullable(ref byte destPtr, MethodTable* typeMT, object? obj)
+        {
+            if (obj == null)
+            {
+                InitValueClass(ref destPtr, typeMT);
+            }
+            else
+            {
+                if (!IsNullableForType(typeMT, RuntimeHelpers.GetMethodTable(obj)))
+                {
+                    // For safety's sake, also allow true nullables to be unboxed normally.
+                    // This should not happen normally, but we want to be robust
+                    if (typeMT == RuntimeHelpers.GetMethodTable(obj))
+                    {
+                        Unsafe.CopyBlockUnaligned(ref destPtr, ref RuntimeHelpers.GetRawData(obj), typeMT->GetNumInstanceFieldBytes());
+                        return;
+                    }
+                    CastHelpers.ThrowInvalidCastException(obj, typeMT);
+                }
+
+                // Set the hasValue field on the Nullable type. It MUST always be placed at the start of the object.
+                *(bool*)destPtr = true;
```

Comment:
This doesn't look correct. The cast is casting the value of the reference into pointer. `Unsafe.As` would be correct.

Reply:
Yeah, that seems wrong. I kept waffling between refs and pointers here. Thanks.